### DISCUSSION TITLE
Incompatibility with Java's versioning scheme

### DISCUSSION THREAD

**Discussion by astellingwerf** [renovatebot/renovate#24159] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v36.67.1

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

The Java versioning scheme is poorly supported in Renovate, because it uses more numbers than a standard semver. Yesterday Adoptium releases version 17.0.8.1+1. That's:

- "major": 17
- "minor": 0
- "security": 8,
- "patch": 1,
- "build": 1,

The Adoptium API does provide a semver equivalent "17.0.8+101" which is useful for comparison and finding the latest, but it can't be used for applications using it's reference to, for example, download artifacts from https://github.com/adoptium/temurin17-binaries/releases. Instead, the field `openjdk_version` is likely what the consumers need.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
I understand what you wrote, but what is the problem needing attention now? For example maybe you can switch to `loose` versioning as it will happily parse/compare a string like `17.0.8.1+1`

**Reply by astellingwerf** [reply]
Does it? I created https://github.com/astellingwerf/renovate-discussion-24159, where I put an out-of-date version for java 11 and 17. For J11 there is a MR, but there's not for J17. The latest version 17 that the DD shows, is `17.0.8+7`

**Reply by astellingwerf** [reply]
I see two problems why this won't work:
- The java-version datasource exposes the mock semver instead of the actual version number. It destroys the significant plusses, and puts plusses where they shouldn't be.
- If the actual version number would be exposed, loose versioning may or may not work, depending on the semver build numbers: When the major.minor.patch numbers are equal (like in `version = "17.0.8+7"`, `other = "17.0.8+101"`), loose versioning will compare the suffix strings, in this case "+7" and "+101", alphabetically, which happens to be reverse numerical order.

**Reply by rarkins** [reply]
So is one approach to have a new datasource called something like `jdk-version` which returns values like `17.0.8.1+1`?

Also I think the loose versioning scheme should be fixed so that it compares +7 and +101 in the way you'd expect, instead of alphabetically

**Reply by astellingwerf** [reply]
We can use a custom/JSONata datasource to get the right version number from the Adoptium API:
```
  "customDatasources": {
    "jdk": {
      "defaultRegistryUrlTemplate": "https://api.adoptium.net/v3/info/release_versions?project=jdk&release_type=ga&sort_method=DATE&sort_order=DESC&show_page_count=true&page_size=100",
      "transformTemplates": [
        "{ \"releases\": $map($.versions, function($v, $i, $a) { { \"version\": $v.openjdk_version} } ) }"
      ]
    }
  }
```
A tricky aspect of that was that I needed to separate the curly braces by spaces to prevent confusion with Moustache templates. Would it make sense to warn folks for that on https://docs.renovatebot.com/modules/datasource/custom/?

I don't know whether there's know sort algorithm for two string with shared prefixes followed by different numbers.

**Reply by yury-komarov-contextlabs** [reply]
Rising up topic

What i have:

renovate.json
```
  "customDatasources": {
    "adoptium": {
      "defaultRegistryUrlTemplate": "https://api.adoptium.net/v3/info/release_versions?project=jdk&release_type=ga&sort_method=DATE&sort_order=DESC&show_page_count=true&page_size=100",
      "transformTemplates": [
        "{ \"releases\": $map($.versions, function($v, $i, $a) { { \"version\": $v.semver} } ) }"
      ]
    }
  },
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["project\\.json"],
      "datasourceTemplate": "adoptium",
      "matchStrings": [
        "datasource=(?<datasource>adoptium) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))? version=(?<currentValue>[^\"]+)"
      ]
    }
...
  
```

project.json
```
...
          {
            "name": "JRE_VERSION_RENOVATE",
            "value": "datasource=adoptium depName=java-jre version=1.8.0_302-b08"
          }
...
```

and output

```
           {
             "deps": [
               {
                 "depName": "java-jre",
                 "currentValue": "1.8.0_302-b08",
                 "datasource": "adoptium",
                 "replaceString": "datasource=adoptium depName=java-jre version=1.8.0_302-b08",
                 "updates": [],
                 "packageName": "java-jre",
                 "versioning": "semver-coerced",
                 "warnings": [],
                 "skipReason": "invalid-config"
               }
             ],
             "matchStrings": [
               "datasource=(?<datasource>adoptium) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))? version=(?<currentValue>[^\"]+)"
             ],
             "datasourceTemplate": "adoptium",
             "packageFile": "project.json"
           },
```

What i'm doing wrong?

**Reply by wiebeck** [reply]
Maybe the `semver-coerced` causes problems detecting the full version. From https://docs.npmjs.com/cli/v6/using-npm/semver#coercion it tries to convert versions to semver by cutting off anything that isn't really expected. Cannot go into details because I'm lacking more experience but we also looked at `semver-coerced` and decided against it for exactly this reason.

You should be able to change it using a `packageRule`.

**Reply by viceice** [reply]
I don't think the java-version datasource is the problem. it simply returns the semver field value from the adoptium API.

probably the semver-coerce versioning can be the issue and you should use semver versioning instead.

**Comment by wiebeck** [comment]
Right now we're facing a slightly different issue. [Adoptium released JDK/JRE 17.0.9+9.1](https://github.com/adoptium/temurin17-binaries/releases/tag/jdk-17.0.9%2B9.1) a few days ago. The Adoptium API responds with

```json
{
    "adopt_build_number": 1,
    "build": 9,
    "major": 17,
    "minor": 0,
    "openjdk_version": "17.0.9+9",
    "security": 9,
    "semver": "17.0.9+9.1"
},
```

...as the very first list entry.

Unfortunately Renovate doesn't pick up this version and provides a PR for it, just for `17.0.9+9`. According to the spec the `+9.1` doesn't validate as "build" and the default versioning also doesn't match as it only expects digits but no dots:

```
"versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
```

We tried to work around this by allowing anything as build:

```
"versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>.+))?(-(?<compatibility>.*))?$"
```

But still, Renovate seems to ignore the `+9.1` release. I have set up a minimal reproduction repository at https://github.com/wiebeck/renovate-java-version/pull/2 where you can take a look at my config.

We're also confused whether this `+9.1` build is just for Windows (unfortunately there is no `windows-x64` build of `17.0.9+9`) or if other platforms will be supported as well as we cannot currently deal with different JRE build versions for different platforms (something that we might need to change) in our product.

**Reply by wiebeck** [reply]
Interestingly version `21.0.1+12.0.LTS` is being found properly by Renovate but not `21.0.1+12.1.LTS`. Take a look at the API output:

```json
{
    "adopt_build_number": 1,
    "build": 9,
    "major": 17,
    "minor": 0,
    "openjdk_version": "17.0.9+9",
    "security": 9,
    "semver": "17.0.9+9.1"
},
{
    "adopt_build_number": 1,
    "build": 12,
    "major": 21,
    "minor": 0,
    "openjdk_version": "21.0.1+12-LTS",
    "optional": "LTS",
    "security": 1,
    "semver": "21.0.1+12.1.LTS"
},
{
    "build": 12,
    "major": 21,
    "minor": 0,
    "openjdk_version": "21.0.1+12-LTS",
    "optional": "LTS",
    "security": 1,
    "semver": "21.0.1+12.0.LTS"
},
{
    "build": 9,
    "major": 17,
    "minor": 0,
    "openjdk_version": "17.0.9+9",
    "security": 9,
    "semver": "17.0.9+9"
},
```

For both the 17 and 21 releases `major`, `minor`, `security` and `build` are identical, just the `semver` differs. It looks like Renovate merges the results with same `maj.min.sec.build` version, so that versions being lower in the list overwrite the earlier ones which would explain that `17.0.9+9.1` and `21.0.1+12.1.LTS` get "lost in translation". ;-) Just a theory though, no idea how to confirm this since I don't know the Renovate source code well enough.

**Reply by yury-komarov-contextlabs** [reply]
Can u share your full example of renovatebot configuration for java-version? Cause for me any tries were unsuccessful :/

**Reply by viceice** [reply]
please change the versioning to semver, your regex versioning is wrong. build can only be an Integer, not float

**Reply by wiebeck** [reply]
Yes, build can only be an integer. I think that's the whole point why `17.0.9+9.1` doesn't qualify as update candidate since the build part here is `9.1` which is simply invalid. We deliberately changed the regex to allow even invalid build parts.

My current config looks like this: I defined a new `customDatasource`:

```json
"adoptium": {
  "defaultRegistryUrlTemplate": "https://api.adoptium.net/v3/info/release_versions?project=jdk&image_type=jre&release_type=ga&sort_method=DATE&sort_order=DESC&page_size=50&page=0",
  "transformTemplates": [
    "{ \"releases\": $map($.versions, function($v, $i, $a) { { \"version\": $v.semver} } ) }"
  ]
}
```

As far as I understand this maps the original `semver` values to something like this:

```json
{
  "releases": [
    {
      "version": "17.0.9+9.1"
    },
    {
      "version": "21.0.1+12.1.LTS"
    },
    {
      "version": "21.0.1+12.0.LTS"
    },
    {
      "version": "17.0.9+9"
    }
  ]
}
```

Interestingly this is basically the exact same as what the `java-version` datasource does:

```ts
const pgRes = await this.http.getJson<AdoptiumJavaResponse>(pgUrl);
return (
  pgRes?.body?.versions?.map(({ semver }) => ({
    version: semver,
  })) ?? null
);
```

Then I configured a `customManager`:

```json
{
  "customType": "regex",
  "description": "Update Java runtime in install4j projects",
  "fileMatch": ["^src/install4j/scope.*\\.install4j"],
  "matchStrings": ["jdkProviderId=\"Adoptium\" release=\"17/jdk-(?<currentValue>.*?)\""],
  "depNameTemplate": "java",
  "versioningTemplate": "semver",
  "datasourceTemplate": "custom.adoptium",
  "packageNameTemplate": "java-jre"
}
```

Unfortunately this only creates update-PRs to `17.0.9+9` and `21.0.1+12.0.LTS` but not `17.0.9+9.1` or `21.0.1+12.1.LTS`. At first I thought this might be because `17.0.9+9.1` isn't a valid semver, so I provided a custom `versioningTemplate` allowing for arbitrary "builds". Still, version `17.0.9+9.1` is not being discovered (and offered) by Renovate. :-/

On the other hand, if the semver really is invalid, then `21.0.1+12.0.LTS` should be invalid as well, right? But this version is being picked up by Renovate. I really don't get why this is the case.

Unfortunately I cannot setup a VS Code dev container (`pnpm build` fails), so I cannot produce a test case for it. But I am sure you can?

**Reply by viceice** [reply]
the basic issue is everything after the `+` is semver metadata and should be ignored when comparing versions. so it's probably filtered as duplicate.

see point 10

https://semver.org/

**Comment by wiebeck** [comment]
My final workaround is a custom datasource slightly different from what has been proposed above:
```json
"java-winx64-latest": {
  "defaultRegistryUrlTemplate": "https://api.adoptium.net/v3/info/release_versions?project=jdk&image_type=jre&release_type=ga&sort_method=DATE&sort_order=DESC&page_size=50&os=windows&architecture=x64",
  "transformTemplates": [
    "{ \"releases\": $map($filter($.versions, function($v) { $v.major = 17 })[0], function($v) { { \"version\": $v.semver } } )[] }"
  ]
},
```
The datasource only fetches versions for a particular os/arch combination and the JSONata transformation filters for the **single latest** version of a particular major version (17 in our case). The resulting list of releases only contains **exactly one entry**. This was the only way I could convince Renovate to offer version **17.0.9+9.1** as update.
